pkg/nodeagent/controller/operatingsystemconfig/reconciler.go

```diff
@@ -0,0 +1,318 @@
+		return reconcile.Result{}, nil
+	}
+
+	log.Info("Applying new or changed files")
+	if err := r.applyChangedFiles(log, oscChanges.files.changed); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed applying changed files: %w", err)
+	}
+
+	log.Info("Applying new or changed units")
+	if err := r.applyChangedUnits(ctx, log, oscChanges.units.changed); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed applying changed units: %w", err)
+	}
+
+	log.Info("Removing no longer needed units")
+	if err := r.removeDeletedUnits(ctx, log, node, oscChanges.units.deleted); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed removing deleted units: %w", err)
+	}
+
+	log.Info("Reloading systemd daemon")
+	if err := r.DBus.DaemonReload(ctx); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed reloading systemd daemon: %w", err)
+	}
+
+	log.Info("Executing unit commands (start/stop)")
+	if err := r.executeUnitCommands(ctx, log, node, oscChanges.units.changed); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed executing unit commands: %w", err)
+	}
+
+	log.Info("Removing no longer needed files")
+	if err := r.removeDeletedFiles(log, oscChanges.files.deleted); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed removing deleted files: %w", err)
+	}
+
+	log.Info("Successfully applied operating system config",
+		"changedFiles", len(oscChanges.files.changed),
+		"deletedFiles", len(oscChanges.files.deleted),
+		"changedUnits", len(oscChanges.units.changed),
+		"deletedUnits", len(oscChanges.units.deleted),
+	)
+
+	log.Info("Persisting current operating system config as 'last-applied' file to the disk", "path", lastAppliedOperatingSystemConfigFilePath)
+	if err := afero.WriteFile(r.FS, lastAppliedOperatingSystemConfigFilePath, oscRaw, 0644); err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to write current OSC to file path %q: %w", lastAppliedOperatingSystemConfigFilePath, err)
+	}
+
+	if node == nil {
+		log.Info("Waiting for Node to get registered by kubelet, requeuing")
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	r.Recorder.Event(node, corev1.EventTypeNormal, "OSCApplied", "Operating system config has been applied successfully")
+
+	patch := client.MergeFrom(node.DeepCopy())
+	metav1.SetMetaDataAnnotation(&node.ObjectMeta, v1beta1constants.LabelWorkerKubernetesVersion, r.Config.KubernetesVersion.String())
+	metav1.SetMetaDataAnnotation(&node.ObjectMeta, executor.AnnotationKeyChecksum, oscChecksum)
+	return reconcile.Result{RequeueAfter: r.Config.SyncPeriod.Duration}, r.Client.Patch(ctx, node, patch)
+}
+
+var etcdSystemdSystem = path.Join("/", "etc", "systemd", "system")
+
+func (r *Reconciler) applyChangedFiles(log logr.Logger, files []extensionsv1alpha1.File) error {
+	tmpDir, err := afero.TempDir(r.FS, "/tmp", "gardener-node-agent-*")
+	if err != nil {
+		return fmt.Errorf("unable to create temporary directory: %w", err)
+	}
+
```

Comment:
```suggestion
	defer func() { utilruntime.HandleError(r.FS.RemoveAll(imageMountDirectory)) }()
```
The temp dir should be removed when we are finished.